### ISSUE TITLE
Do no compile library items that are not used

### ISSUE THREAD

**Issue by mohammadfawaz** [FuelLabs/sway#2873]
We could do this in two steps:
1. Do not compile _whole libraries_ that are not needed (or transitively needed) by the entry program.
2. Do not compile library _items_ that are not needed (or transitively needed) by the entry program.

(2) will become easier when the collection context work is complete (I think).

**Comment by otrho** [comment]
(1) is already kinda handled by `Forc.toml` isn't it?  We only compile the libraries in there before providing them to the entry.

**Comment by mohammadfawaz** [comment]
Yeah thinking about "sub-libraries" of the standard library here as an example. If the top level program does not need `assert` anywhere, even transitively, then don't compile the `assert` library.
